pkg/embeddedcluster/monitor.go

```diff
@@ -0,0 +1,139 @@
+package embeddedcluster
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/replicatedhq/embedded-cluster-operator/api/v1beta1"
+	apptypes "github.com/replicatedhq/kots/pkg/app/types"
+	statetypes "github.com/replicatedhq/kots/pkg/appstate/types"
+	"github.com/replicatedhq/kots/pkg/k8sutil"
+	"github.com/replicatedhq/kots/pkg/logger"
+	"github.com/replicatedhq/kots/pkg/store"
+	storetypes "github.com/replicatedhq/kots/pkg/store/types"
+	"k8s.io/client-go/kubernetes"
+)
+
+// StartInstallationMonitor starts a goroutine that monitors the embedded cluster installation
+// and starts the upgrade process if necessary.
+func StartInstallationMonitor(ctx context.Context) error {
+	clientset, err := k8sutil.GetClientset()
+	if err != nil {
+		return fmt.Errorf("failed to get kubeclient: %w", err)
+	}
+	if isembedded, err := IsEmbeddedCluster(clientset); err != nil {
+		return fmt.Errorf("failed to check if embedded: %w", err)
+	} else if !isembedded {
+		return nil
+	}
+	mon := monitor{
+		store:     store.GetStore(),
+		clientset: clientset,
+	}
+	go mon.start(ctx)
+	return nil
+}
+
+// monitor is a struct that groups all methods needed to monitor the embedded cluster installation.
+type monitor struct {
+	store     store.Store
+	clientset *kubernetes.Clientset
+}
+
+// getApp returns the app deployed on top of the embedded cluster.
+func (m *monitor) getApp(ctx context.Context) (*apptypes.App, error) {
+	apps, err := m.store.ListInstalledApps()
+	if err != nil {
+		return nil, fmt.Errorf("failed to list installed apps: %w", err)
+	} else if len(apps) == 0 {
+		return nil, nil
+	}
+	return apps[0], nil
+}
+
+// maybeStartUpgrade checks if the embedded cluster is in a state that requires an upgrade. If so,
+// it starts the upgrade process. We only start an upgrade if the following conditions are met:
+// - We have an app deployed on top of the embedded cluster.
+// - The deployed app version is in ready state.
```

Comment:
I think we specifically do NOT want to wait for the app to be in a ready state - and in fact might want to deploy the cluster update _even if the app could not be applied to the cluster_.

This is for situations where an app version uses a resource introduced in k8s 1.X, and the cluster is currently running 1.X-1 (among other possibilities)